Follow smart contract

darcys22 opened the pull request.

darcys22 committed:
Load ethyl library and basic sepolia parameters

darcys22 committed:
adds a L2 tracker to core, which checks the smart contract every 10 secs for it state and maintains the history so block verifiers can check

darcys22 committed:
hardforks for devnet

darcys22 committed:
rpc commands needing updating to new format

darcys22 committed:
wallet2 updates

darcys22 committed:
local devnet updates to latest hardfork

darcys22 committed:
l2 tracker data

darcys22 committed:
hardfork devnet to HF20

darcys22 committed:
more block header info

darcys22 committed:
rewards actually reads

darcys22 committed:
update ethyl to include height in eth calls

darcys22 committed:
l2 tracker keeps state up to date

darcys22 committed:
checks state before including it in block

darcys22 committed:
start on BLS signing

darcys22 committed:
update local devnet codes

darcys22 committed:
bls library

darcys22 committed:
bls signing

darcys22 committed:
use connection for omq

darcys22 committed:
rpc call for bls

darcys22 committed:
bls signing not working

darcys22 committed:
aggregate signing works

darcys22 committed:
folders for local devnet to exclude changing ip addresses

darcys22 committed:
merkle tree creator

darcys22 committed:
merkle tree from batching rewards

darcys22 committed:
rpc request for merkle tree

darcys22 committed:
update to latest ethyl

darcys22 committed:
cli command to get link to claim rewards

darcys22 committed:
remove coinbase rewards

darcys22 committed:
delay hardfork in devnet cause no longer getting rewards

darcys22 committed:
rpc request to get all of the bls pubkeys

darcys22 committed:
silence warnings

darcys22 committed:
Adds a new ethereum transaction to update nodes about eth address

This will fill in an item on the service_node_list contributors struct
and also add to a mapping in the batching database. So if you want to
query for who to pay it will be able to return an ethereum address to
payout.

darcys22 committed:
put query for nodes into its own function

darcys22 committed:
epee cstdint

darcys22 committed:
new transaction types for ethereum

darcys22 committed:
extract logs from block, get state changes as transactions in l2 tracker

darcys22 committed:
transactions added to block

darcys22 committed:
sends transaction to contract

darcys22 committed:
local devnet will spin up a node and submit a register transaction through ethereum

darcys22 committed:
flags for setting provider

darcys22 committed:
submit the transaction

darcys22 committed:
transaction gets included in a block

darcys22 committed:
node gets added to sn list from eth transaction

darcys22 committed:
Ethereum service node rewards recorded to an ethereum address

darcys22 committed:
Unlock Service Node - Ethereum Transaction

darcys22 committed:
Claim rewards functioning

darcys22 committed:
bls rpc calls for exiting and liquidation

darcys22 committed:
exiting adds to the batching database

Commit:
Update static build ZLIB url to Github

Archived versions of ZLIB are pruned from the official zlib
website. These versions are preserved in the Github which
make it more reliable to use a source for static building.

Commit:
Add libgmp to requirements for ethyl dependency

darcys22 committed:
multiple contributors data, registration signature validation

darcys22 committed:
change zero length arrays to nullptr

darcys22 committed:
core tests passing

Doy-lee commented on src/blockchain_db/lmdb/db_lmdb.cpp:
```diff
@@ -1020,7 +1020,7 @@ uint64_t BlockchainLMDB::add_transaction_data(
     } else if (result != MDB_NOTFOUND) {
         throw1(DB_ERROR(lmdb_error(
-                "Error checking if tx index exists for tx hash " + tools::type_to_hex(tx_hash) +
+                "error checking if tx index exists for tx hash " + tools::type_to_hex(tx_hash) +
```
It looks like you did a search and replace here and there's a bunch of churn here that's probably not needed in these capitalisation changes?

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/c222dcbfb45060237b4deb776211866e553e1827

Doy-lee commented on src/CMakeLists.txt:
```diff
@@ -31,7 +31,8 @@
 if (NOT (MINGW OR APPLE OR FREEBSD OR OPENBSD OR DRAGONFLY))
-  add_compile_options("${WARNINGS_AS_ERRORS_FLAG}") # applies only to targets that follow
+  message(STATUS "Setting warnings as errors")
+  #add_compile_options("${WARNINGS_AS_ERRORS_FLAG}") # applies only to targets that follow
```
Temporary change? Ready to be restored?

Doy-lee commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -176,6 +176,27 @@ void BlockchainSQLite::upgrade_schema() {
+    const auto eth_mapping_table_count = prepared_get<int64_t>(
+            "SELECT COUNT(*) FROM sqlite_master WHERE type='table' AND "
+            "name='eth_mapping';");
+    if (!eth_mapping_table_count == 0) {
```
Is this meant to be, `if (eth_mapping_table_count == 0) {`?

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/e43a91c1136de44af891e89caf435239147e876a

Doy-lee commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -176,6 +176,27 @@ void BlockchainSQLite::upgrade_schema() {
+        log::info(logcat, "Adding eth mapping table to batching db");
+        auto& netconf = get_config(m_nettype);
+        SQLite::Transaction transaction{db, SQLite::TransactionBehavior::IMMEDIATE};
+        db.exec(fmt::format(
```
We can just inline the raw-string literal since we don't have any variable substitution

`db.exec(R"(...)");`

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/ffc47b4b07397524fed1f567f8b73f6b1d879e65

Doy-lee commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -288,13 +309,28 @@ void BlockchainSQLite::blockchain_detached(uint64_t new_height) {
+            " ON CONFLICT (address) DO UPDATE SET amount = amount + excluded.amount"
+            );
+    db::exec_query(update_address, tools::type_to_hex(eth_address), oxen_address);
+    return true;
```
The bool return here probably wants to be

```
bool result = db::exec_query(...) > 0; // Number of rows affected
return result;
```

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/954fd4c7a27fcd2025a526127a79e79ee6b8019a

Doy-lee commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -307,7 +343,7 @@ bool BlockchainSQLite::add_sn_rewards(const std::vector<cryptonote::batch_sn_pay
                 static_cast<int>(payment.address_info.address.modulus(netconf.BATCHING_INTERVAL));
         auto amt = static_cast<int64_t>(payment.amount);
-        const auto& address_str = get_address_str(payment.address_info.address);
+        const auto& address_str = get_address_str(payment);
```
`const auto address_str` since the function returns by value.

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/dcc2fc3dfba3f5d0c0166810d46d892ae5261152

Doy-lee commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -327,7 +363,7 @@ bool BlockchainSQLite::subtract_sn_rewards(
 
     for (auto& payment : payments) {
-        const auto& address_str = get_address_str(payment.address_info.address);
+        const auto& address_str = get_address_str(payment);
```
`const auto address_str` since the function returns by value.

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/dcc2fc3dfba3f5d0c0166810d46d892ae5261152

Doy-lee commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -587,6 +649,35 @@ bool BlockchainSQLite::add_block(
+
+        payments.emplace_back(eth_address, amount);
+
+        if (!add_sn_rewards(payments))
```
This function can't actually fail, always returns true or throws an exception. We either make it failable by checking the `db::exec` call or forgo it and remove bool from the function definition.

To me it looks like it'd be something has seriously gone wrong if this fails, so it's worth making the API return the boolean correctly.

Doy-lee commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -704,7 +795,7 @@ bool BlockchainSQLite::save_payments(
 
     for (const auto& payment : paid_amounts) {
-        const auto& address_str = get_address_str(payment.address_info.address);
+        const auto& address_str = get_address_str(payment);
```
`const auto address_str` since the function returns by value.

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/dcc2fc3dfba3f5d0c0166810d46d892ae5261152

Doy-lee commented on src/bls/bls_aggregator.cpp:
```diff
@@ -0,0 +1,178 @@
+                    // Data contains -> status, address, amount, height, bls_pubkey, signed message, signature
+                    uint64_t current_amount = std::stoull(data[2]);
+                    uint64_t current_height = std::stoull(data[3]);
+                    if (!initial_data_set) {
```
Race condition on this variable

Doy-lee replied:
Also since this is unsafe, this has a pretty crap implication on the current approach. We can't necessarily trust the initial data as the reference. We more likely have to actually aggregate the responses and use the majority.

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/7556677cca601e3dd3627042ce631f5ff2f9dc52 

As for supermajority check, can be saved for future work.

Doy-lee commented on src/bls/bls_aggregator.cpp:
```diff
@@ -0,0 +1,178 @@
+            if (success) {
+                if (data[0] == "200") {
+
+                    // Data contains -> status, address, amount, height, bls_pubkey, signed message, signature
```
These OMQ request in general are untrusted, so a lot of checking needs to be done here. On that vein, `stoull` is unsafe and needs to be wrapped in try-catch to avoid blowing up the node.

But note that special attention will need to be paid to all these RPC requests because it's all untrusted data afaict.

A sanitised data change can come after the integration branch.

Doy-lee commented on src/bls/bls_aggregator.cpp:
```diff
@@ -0,0 +1,178 @@
+                if (data[0] == "200") {
+
+                    // Data contains -> status, bls_pubkey (signer), bls_pubkey (node being removed), signed message, signature
+                    if (!initial_data_set) {
```
Race condition on initial data set

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/7556677cca601e3dd3627042ce631f5ff2f9dc52

Doy-lee commented on src/bls/bls_aggregator.cpp:
```diff
@@ -0,0 +1,178 @@
+                if (data[0] == "200") {
+
+                    // Data contains -> status, bls_pubkey (signer), bls_pubkey (node being removed), signed message, signature
+                    if (!initial_data_set) {
```
Race condition on initial data set

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/7556677cca601e3dd3627042ce631f5ff2f9dc52

Doy-lee commented on src/bls/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+  PUBLIC
+    bls::bls256
+  PRIVATE
+    cncrypto
```
Some of these need to be public since we use the definitions and their prototypes in the header of our BLS module like ServiceNodeList and OxenMQ

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/77fcd51c935fd5cb4e891e6cc47d00a2a3e1fd81

Doy-lee commented on src/bls/bls_aggregator.h:
```diff
@@ -0,0 +1,156 @@
+
+public:
+    BLSAggregator(service_nodes::service_node_list& _snl, std::shared_ptr<oxenmq::OxenMQ> _omq, std::shared_ptr<BLSSigner> _bls_signer);
+    ~BLSAggregator();
```
We can remove the destructor since it doesn't do anything

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/571d170b7e0fa2b23883713d59fbaa747b5c82c4

Doy-lee commented on src/cryptonote_core/blockchain.h:
```diff
@@ -1158,6 +1163,10 @@ class Blockchain {
 
+    void add_ethereum_transactions_to_tx_pool();
+
+    std::shared_ptr<L2Tracker> m_l2_tracker;
```
Shared pointers seem to be liberally used in the codebase which is very suspicious. From a glance didn't seem like we have a shared pointer situation. Once shared pointers are used and shared around, it's hard to reign in and maintain a strong understanding of life-times and consequently confidence in the memory imprint of the codebase. Recommend to downgrade to unique pointer, or, better, the applications of these smart pointers don't seem to be dealing with objects that need heap storage. It looks like pointers are being used as a way to avoid initialisation of a constructor (because the constructor locks in a design, e.g. L2 starting a thread on construction) should be avoided and instead done using methods to state transition the object.

Foot note here is that typically every memory allocation is a potential call into the kernel which is a ring transition. It's slow, subject to OS errors/faults and _can_ fail. This is avoided entirely if it's stack allocated.

Doy-lee commented on src/bls/bls_aggregator.h:
```diff
@@ -0,0 +1,156 @@
+#pragma GCC diagnostic ignored "-Wshadow"
+#pragma GCC diagnostic ignored "-Wconversion"
+#pragma GCC diagnostic ignored "-Wsign-conversion"
+#include <bls/bls.hpp>
```
Looks like it can be moved into the implementation file

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/110571749ec2800b6ea4ca82c02d669e094197b4

Doy-lee commented on src/bls/bls_aggregator.cpp:
```diff
@@ -0,0 +1,178 @@
+
+    processNodes(
+        "bls.get_reward_balance",
+        [this, &aggSig, &signers, &signers_mutex, &lower_eth_address, &amount, &height, &signed_message, &initial_data_set](bool success, std::vector<std::string> data) {
```
The callback should be passed in the node IP and public key so that we can log failures with more descriptive information which will be valuable when trying to post-mortem debug logs

Doy-lee replied:
Callback can also take `const std::vector<std::string>& data`

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/065de38e24f0dc9159af452b0201b760f37f33b9

Doy-lee commented on src/bls/bls_aggregator.cpp:
```diff
@@ -0,0 +1,178 @@
+        [this, &pubkeys, &pubkeys_mutex](bool success, std::vector<std::string> data) {
+            if (success) {
+                std::lock_guard<std::mutex> lock(pubkeys_mutex);
+                pubkeys.emplace_back(data[0], data[1]);
```
`data` needs to be bound check, can't trust the incoming data. Similar for the rest of the requests, some more bounds checking is needed.

A sanitised data change can come after the integration branch.

Doy-lee commented on src/bls/bls_aggregator.h:
```diff
@@ -0,0 +1,156 @@
+    // `callback` will be called to process their reply and after everyone has been received it will then call
+    // `postProcess`
+    template <typename Callback, typename PostProcess>
+    void processNodes(const std::string& request_name, Callback callback, PostProcess postProcess, const std::optional<std::string>& message = std::nullopt) {
```
`processNodes` doesn't need to use template `Callback`/`PostProcess` because the code invoking them is hardcoded to have a specific prototype `callback(/*bool*/ success, /*std::vector*/ data)`. This then means we can move implementation into the .cpp file

It's tempting to have a generic-catch all approach but that tends to complicate things early on (e.g. the templating, or suggestion that this function is used with multiple callback types), rather avoid it until we actually have a problem that requires that approach.

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/78f37f832f1f3ad181c91e81957fab3441a6be49

Doy-lee commented on src/bls/bls_aggregator.h:
```diff
@@ -0,0 +1,156 @@
+private:
+    // Goes out to the nodes on the network and makes oxenmq requests to all of them, when getting the reply
+    // `callback` will be called to process their reply and after everyone has been received it will then call
+    // `postProcess`
```
`processNodes` probably needs a supermajority success rate before it returns success. It should return whether or not the output can be trusted or not (e.g. >= supermajority, trust-worth else untrusted response)

Can be fixed as part of future work.

Doy-lee commented on src/bls/bls_aggregator.h:
```diff
@@ -0,0 +1,156 @@
+                ++active_connections;
+            }
+            
+            auto conn = omq->connect_remote(
```
There's a `connect_sn` function that we probably want to use instead of `connect_remote`. This uses a SN pubkey so it'd avoid having to pull in boost::asio for IPv4 creation and defer to the library itself.

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/b1326d5567d3090bece85a100e73399e252a2ad0

Doy-lee commented on src/bls/bls_aggregator.h:
```diff
@@ -0,0 +1,156 @@
+        size_t active_connections = 0;
+        const size_t MAX_CONNECTIONS = 900;
+
+        // TODO sean, change this so instead of using an iterator do a for_each_service_node_info_and proof and pass a lambda
```
re: The todo

`processNodes` takes unsafe iterator to service-node list. IIRC the rpc server is run on a separate thread so a lock needs to be taken on the core add block process for this to be safe. My suggestion is a gather then unlock the list and work off thread-locally owned data. I suggest this over the lambda because throughput will be higher than `for_each_service_node..` since business logic is separated from the gather step.

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/03e0379c382eaa51c535e2464a7b06c4931beee6

Doy-lee commented on src/bls/bls_aggregator.h:
```diff
@@ -0,0 +1,156 @@
+            return active_connections == 0;
+        });
+        
+        postProcess();
```
I suggest we remove `postProcess` until we have some functionality that needs it

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/78f37f832f1f3ad181c91e81957fab3441a6be49

darcys22 replied:
yeah this got changed and is just dead code. I originally thought the node would itself need to sign separately.

Doy-lee commented on src/bls/bls_signer.cpp:
```diff
@@ -0,0 +1,104 @@
+}
+
+void BLSSigner::initCurve() {
+    // Initialize parameters for BN256 curve, this has a different name in our library
```
Are we sure we need to do this initialisation? `herumi/bls` has this in their readme, e.g. BLS_ETH is sufficient to configure the library, although that might not be handling everything we're trying to do.

```
If you want to use the same parameters as Ethereum 2.0, just define BLS_ETH. If you want to use mcl/bls without BLS_ETH, then check the following settings.
```

Because usually these kind of one-time init patterns on libraries that don't take in a context modify some static state which might not be thread-safe in which case this can result in very difficult to detect behaviour in the program.

If there's a specific need to initialise library configuration, I'd place this as early as possible in the program, in blockchain init or the `main` entry point.

darcys22 replied:
Yeah, we dont want to use the same parameters as eth2.0. We have a different curve. We do need the BLS_ETH to put the public key on the G1 curve though. So we need both

Doy-lee commented on src/bls/bls_signer.cpp:
```diff
@@ -0,0 +1,104 @@
+    bool b;
+    mcl::bn::mapToG1(&b, gen, 1);
+    blsPublicKey publicKey;
+    publicKey.v = *reinterpret_cast<const mclBnG1*>(&gen);
```
This would need to use memcpy to be C++ compliant

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/8aec3d1e30866a20bacd149d6d97a8756200304b

Doy-lee commented on src/bls/bls_signer.cpp:
```diff
@@ -0,0 +1,104 @@
+std::string BLSSigner::buildTag(const std::string_view& baseTag, uint32_t chainID, const std::string& contractAddress) {
+    // Check if contractAddress starts with "0x" prefix
+    std::string contractAddressOutput = contractAddress;
+    if (contractAddressOutput.substr(0, 2) == "0x")
```
A safer and non-allocating alternative to this is a having a `trimPrefix()` that does _not_ throw and returns a string_view. Although at the moment I am able to prove that contractAddress will always be valid (seems we always pull from some statically declared strings) moving forward this might not be as easy to determine or assert as the codebase changes.

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/8aec3d1e30866a20bacd149d6d97a8756200304b

Doy-lee commented on src/bls/bls_signer.cpp:
```diff
@@ -0,0 +1,104 @@
+    return publicKey;
+}
+
+std::array<unsigned char, 32> BLSSigner::hash(std::string in) {
```
Should take in string_view

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/8aec3d1e30866a20bacd149d6d97a8756200304b

Doy-lee commented on src/bls/bls_signer.cpp:
```diff
@@ -0,0 +1,104 @@
+    return utils::toHexString(utils::hash(concatenatedTag));
+}
+
+std::string BLSSigner::buildTag(const std::string_view& baseTag) {
```
We can remove the reference and const qualifier on `string_view`

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/8aec3d1e30866a20bacd149d6d97a8756200304b

Doy-lee commented on src/bls/bls_signer.cpp:
```diff
@@ -0,0 +1,104 @@
+    blsSetGeneratorOfPublicKey(&publicKey);
+}
+
+std::string BLSSigner::buildTag(const std::string_view& baseTag, uint32_t chainID, const std::string& contractAddress) {
```
Can remove the const and reference qualifier on `string_view`

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/8aec3d1e30866a20bacd149d6d97a8756200304b

Doy-lee commented on src/bls/bls_signer.cpp:
```diff
@@ -0,0 +1,104 @@
+    std::array<unsigned char, 32> serialized_hash;
+    uint8_t *hdst = serialized_hash.data();
+    mclSize serializedSignatureSize = 32;
+    if (x.serialize(hdst, serializedSignatureSize, mcl::IoSerialize | mcl::IoBigEndian) == 0)
```
`if (x.serialize(hdst, hdst.max_size(), mcl::IoSerialize | mcl::IoBigEndian) == 0)`

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/8aec3d1e30866a20bacd149d6d97a8756200304b

Doy-lee commented on src/bls/bls_signer.h:
```diff
@@ -0,0 +1,56 @@
+    static std::array<unsigned char, 32> hash(std::string in);
+    static std::array<unsigned char, 32> hashModulus(std::string message);
+
+    static constexpr std::string_view proofOfPossessionTag = "BLS_SIG_TRYANDINCREMENT_POP"sv;
```
`static constexpr inline` Can be used in C++17 otherwise some compiler settings might complain about this

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/4166714377bd1d30734b5831647439d641d9c7dd

Doy-lee commented on src/bls/bls_utils.cpp:
```diff
@@ -0,0 +1,39 @@
+    std::vector<unsigned char> serialized_signature(serializedSignatureSize*4);
+    uint8_t *dst = serialized_signature.data();
+    const blsSignature* blssig = sig.getPtr();  
+    const mcl::bn::G2* g2Point = reinterpret_cast<const mcl::bn::G2*>(&blssig->v);
```
Can reuse from `ethyl`

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/73b8e5533779456b0eca7a072c2937404e4779ed

Doy-lee commented on src/crypto/base.h:
```diff
@@ -18,8 +18,8 @@ constexpr T null{};
-template <size_t Bytes, bool MemcmpHashHex = false>
-struct alignas(size_t) bytes {
+template <size_t Bytes, bool MemcmpHashHex = false, typename AlignAs = size_t>
+struct alignas(AlignAs) bytes {
```
`struct alignas(AlignAs) bytes` we can avoid the templating on AlignAs to simplify compile complexity. The recipient structs should instead specify alignment in their declaration for explicitness, e.g., `struct alignas(uint32_t) eth_address : bytes<32>`

Doy-lee replied:
I tried this but I deemed it not worth it. A bunch of code does `reinterpret_cast<size_t*>(&bytes);` on various objects using this class and tracking them all down and ensuring there are not regression is not easy to test right now.

Doy-lee commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -825,6 +834,10 @@ block Blockchain::pop_block_from_blockchain(bool pop_batching_rewards = true) {
 
     // return transactions from popped block to the tx_pool
     size_t pruned = 0;
+    std::shared_ptr<TransactionReviewSession> ethereum_transaction_review_session;
```
Shared pointer is pretty heavy-handed for this object's lifetime, we can scope it down to a stack allocated variable which will make lifetimes a lot simpler and more performant.

Doy-lee commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -1925,6 +1947,66 @@ bool Blockchain::create_next_pulse_block_template(
+    tx_extra_field field;
+
+    for (const auto& tx_variant : eth_transactions) {
+        transaction tx;
```
As discussed in offline meeting earlier there's TX ordering problem here. This seems like a sizable ticket item to work on that we can save it for after we initiate the integration branch.

The easy solution is to just bundle all these up into one TX extra. We need to order the TX's from the L2 tracker. This can be done simply by doing a sort based on `(block_index, tx_index)`.

Doy-lee commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -1071,6 +1093,73 @@ void core::init_oxenmq(const boost::program_options::variables_map& vm) {
+                        "(received " +
+                        std::to_string(m.data.size()) + ")");
+                std::string eth_address = std::string(m.data[0]);
+                if (eth_address.substr(0, 2) != "0x") {
```
This I believe is untrusted input, we'll have to check it before we work with it, `trimPrefix` could be used here to prevent this from blowing up.

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/a80490fe7a4cf222a9dc99bbd6f762b5e846fc24

Doy-lee commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -1071,6 +1093,73 @@ void core::init_oxenmq(const boost::program_options::variables_map& vm) {
+                std::string encoded_message = "0x" + m_bls_signer->buildTag(m_bls_signer->rewardTag) + utils::padToNBytes(eth_address.substr(2), 20, utils::PaddingDirection::LEFT) + utils::padTo32Bytes(utils::decimalToHex(amount), utils::PaddingDirection::LEFT);
+                const auto h = m_bls_signer->hash(encoded_message);
+                // Returns status, address, amount, height, bls_pubkey, signed message, signature
+                m.send_reply("200", m.data[0], std::to_string(amount), std::to_string(batchdb_height), m_bls_signer->getPublicKeyHex(), encoded_message, m_bls_signer->signHash(h).getStr());
```
The OMQ responses should send back binary to be optimal and minimise the bandwidth necessary to invoke calls. I'm relegating this to work that can be done after the initial integration branch.

Doy-lee commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -2669,6 +2765,51 @@ core::get_service_node_blacklisted_key_images() const {
+        auto& key = service_node_pubkeys.emplace_back();
+        tools::hex_to_type(node.first, key);
+        const auto infos = m_service_node_list.get_service_node_list_state(service_node_pubkeys);
+        bls_pubkeys_and_amounts.emplace_back(node.second, infos[0].info->total_contributed);
```
Untrusted input can blow up node. Fix can be relegated to after initial integration branch.

Doy-lee commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -3918,6 +4203,27 @@ crypto::public_key service_node_list::get_random_pubkey() {
 
+// Function to get an iterator for the first service node info
+service_nodes::service_nodes_infos_t::iterator service_node_list::get_first_pubkey_iterator() {
+    std::lock_guard lock{m_sn_mutex};
```
If a lock is being contended on this API at this point, I think it's already too late, very likely your iterator will get invalidated. We can keep the API but it needs a massive footnote that a lock must be held by whoever's request the iterator from the list.

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/5c2bebbd276140824dafc05553c90a757c30b7bb

I removed the API because I've achieved what we needed to do more optimally without needing to expose iterators. We can re-add in the future if a use-case requires.

Doy-lee commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -3970,6 +4276,31 @@ std::string service_node_list::remote_lookup(std::string_view xpk) {
 }
 
+crypto::public_key service_node_list::bls_public_key_lookup(crypto::bls_public_key bls_key) {
+    std::vector<crypto::public_key> sn_pks;
```
The following sequence of calls is very expensive and we should avoid it. Since we're an internal function we can iterate over the service nodes directly

```cpp
std::lock_guard lock{m_sn_mutex};
for (auto sn_info : m_state.service_nodes_infos) {
    // Do check with it->second.bls_public_key ...
}
```

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/5c2bebbd276140824dafc05553c90a757c30b7bb

Doy-lee commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -3970,6 +4276,31 @@ std::string service_node_list::remote_lookup(std::string_view xpk) {
     return "tcp://" + epee::string_tools::get_ip_string_from_int32(ip) + ":" + std::to_string(port);
 }
 
+crypto::public_key service_node_list::bls_public_key_lookup(crypto::bls_public_key bls_key) {
```
Can pass key by const ref. This function can be `const` as well.

Doy-lee replied:
Sorted in https://github.com/oxen-io/oxen-core/commit/5c2bebbd276140824dafc05553c90a757c30b7bb

Doy-lee commented on src/l2_tracker/l2_tracker.cpp:
```diff
@@ -0,0 +1,276 @@
+
+void L2Tracker::update_state() {
+    try {
+        State new_state(rewards_contract->State());
```
This is all threaded so the entire public API will need to be thread safe. In particular since we check for missing heights after committing the latest height there's a real risk that the application might request the heights inbetween and read an empty state implying there're no L2 state changes.

In general I'd recommend updating in order just for consistency. We can achieve that 

```cpp
State latest_state = rewards_contract->State();
u64 init_height    = std::min(state_history.empty() ? latest_state.height : state_history.back().height + 1);
for (u64 height_it = init_height; height_it < latest_state.height; height_it++) {
	// ... do work
}
```

Doy-lee replied:
Anyway this piece of the puzzle is going to go through a deeper look so how important it is to address it now might not be the top priority.

Doy-lee commented on src/wallet/node_rpc_proxy.cpp:
```diff
@@ -113,10 +113,14 @@ bool NodeRPCProxy::get_info() const {
+            auto it_immutable_height = res.find("immutable_height");
+            if (it_immutable_height != res.end())
+                m_immutable_height = res.at("immutable_height").get<uint64_t>();
+            else
```
The else branch is missing some code?

Doy-lee commented on src/l2_tracker/rewards_contract.cpp:
```diff
@@ -0,0 +1,153 @@
+            // address is 32 bytes , pubkey is 64 bytes and serviceNodePubkey is 64 bytes
+            //
+            // The address is in 32 bytes, but actually only uses 20 bytes and the first 12 are padding
+            int pos = 2; // Start after the 0x prefix
```
This is untrusted data, sanitising it can come after integration branch is merged in.

darcys22 committed:
unit tests passing

darcys22 committed:
bls and ethyl librarys to https instead off ssl

darcys22 committed:
update ethyl to latest

darcys22 committed:
update oxen encoding

darcys22 committed:
add logs to ethyl

darcys22 committed:
update oxen-mq

Doy-lee committed:
Remove oxen-encoding, use oxen-mq/external/oxen-encoding

We can avoid keeping 2 copies of the same library since oxen-encoding in oxen-mq
is exposed as a public library allowing oxen-core to interface with
oxen-encoding via oxen-mq.

This simplifies the build and avoids any need to synchronise the
dependencies together (e.g. oxen-encoding was updated to C++20 but
oxen-mq was still on the C++17 branch). We can move up to C++20 in
unison when oxen-mq is updated to C++20 (and hence avoid any
incompatibility issues).

Doy-lee committed:
Undo unneeded churn on error strings to minimise PR changes

Doy-lee committed:
Fix eth_mapping_table_count typo adding not operator on equality

Doy-lee committed:
Avoid fmt::format when raw string literal suffices

Doy-lee committed:
Calculate a return value for BlockchainSQLite::update_sn_rewards_address

Doy-lee committed:
Fix converting of lvalue to 'const& string' from get_address_str

Doy-lee committed:
Interim fix `initial_data_set` race by locking in BLS agg

Also since nodes are untrusted this has a pretty big implication on the
current approach. We can't necessarily trust the initial data as the
reference. We more likely have to actually aggregate the responses and
use the majority. This should be fix in a polish-pass of this BLS
integration work.

Doy-lee committed:
Remove unused BLSaggregator destructor

Doy-lee committed:
Move BLS headers into impl file for BLSAggregator

darcys22 committed:
save bls key to file and load on startup

Doy-lee committed:
Move BLSAggregator::processNodes into impl file by removing templated callback

Doy-lee committed:
Add addressing info to processNodes callback for diagnostics

Doy-lee committed:
Use  to talk to nodes in BLSAggregator

Doy-lee committed:
Avoid expensive lock/race when iterating the SNs for BLS aggregation

darcys22 committed:
Merge pull request #1 from Doy-lee/doyle-follow-smart-contract

Sort out dependency conflict `external/oxen-encoding` vs `external/oxen-mq/external/oxen-encoding`

darcys22 committed:
start of following pool

Doy-lee committed:
Take in string_view's where possible in bls_signer

Doy-lee committed:
Make strings inline, remove unneeded sv suffix

Doy-lee committed:
Use UB safe impl of BLS primitive conversions, simplify header

Doy-lee committed:
Reuse string helpers in string tools for BLS OMQ input

Doy-lee committed:
Avoid multiple locks and spurious heap allocs in bls_public_key_lookup

Doy-lee committed:
Make some targets public due to header usage

Doy-lee committed:
Use the same techniques to load BLS keys as we do Ed25519

Prevents rewriting a bunch of the same code multiple times. More importantly, in
our Ed25519 code, when we load keys, we ensure we

- Move the buffer with secret data out of the stream
- Deserialize into the struct
- Secure memory wipe of the buffer

Which is a bit more rigorous process for dealing with secrets that is
important not to omit in the BLS process as well.

Doy-lee commented on src/bls/bls_signer.cpp:
```diff
@@ -0,0 +1,109 @@
+    contractAddress = config.ETHEREUM_REWARDS_CONTRACT;
+    // This init function generates a secret key calling blsSecretKeySetByCSPRNG
+    secretKey.init();
+    if (fs::exists(key_filepath)) {
```
There are some helper functions in the codebase for loading in keys which I've updated this code to use which has a more paranoid/rigorous process for dealing with secrets.

Updated at https://github.com/oxen-io/oxen-core/commit/82d585b194b4c026d4cbf2d6f6eec62dc6a07240

Doy-lee committed:
Use correct sqlite target in core_tests to defer to static/sys deps accordingly

darcys22 committed:
oxen-mq to stable

Doy-lee committed:
Bad assertion in json_archive, stack is +=1 for each scope not 2

Doy-lee committed:
Create the .oxen folder before we try create keys

Creation of keys fail if the folder doesn't exist beforehand.

Doy-lee committed:
Correctly de/serialise the BLS key as per API

Doy-lee committed:
Fix unused else branch of immutable_height check

darcys22 committed:
block rewards pulled from rewards pool smart contract

Doy-lee committed:
Allow specifying of binpath from CLI for local devnet sn script

Doy-lee committed:
Update the smart contract hashes to match eth-sn-contracts @ 0a223380813

Doy-lee committed:
Uncomment else if branch prevent service node leave requests from executing

Doy-lee committed:
Switch oxen-mq to stable branch

darcys22 committed:
Merge branch 'follow-smart-contract-pr-review' into follow-smart-contract

Doy-lee committed:
Fix block reward verify fmt log specifying incorrect number of args

Doy-lee committed:
Pre-empt block reward to 0 in chaingen tests when ETH_BLS is enabled

Doy-lee committed:
Add some basic guards around L2 tracker that is use concurrently

Doy-lee committed:
Get ethyl  thread-safe fix & getLogs failing

darcys22 committed:
Merge pull request #2 from Doy-lee/doyle-follow-smart-contract

Fix bad fmt log args, make L2 tracker/ethyl provider thread-safe

jagerman committed:
Unbreak oxen-encoding dependency when oxenmq is installed on the system

The change in c5c5ea071 broke compilation when liboxenmq is installed on
the system: we find liboxenmq via system dep and so never include the
oxen-mq subdirectory, which means the oxenc::oxenc target never gets set
up.
